src/psp/guest_request.rs

```diff
@@ -0,0 +1,531 @@
+            unsafe { EVP_CIPHER_CTX_free(self.ctx_resp) };
+        }
+    }
+
+    fn encrypt_request(&mut self, msg_type: u8, plaintext: &[u8]) -> bool {
+        let msg_seqno: u64 = match seqno_last_used().checked_add(1) {
+            Some(v) => v,
+            None => {
+                prints!("ERROR: Request sequence number overflow\n");
+                return false;
+            }
+        };
+        let aad_len: usize =
+            core::mem::size_of::<SnpGuestRequestMsgHdr>() - offset_of!(SnpGuestRequestMsgHdr, algo);
+        let mut iv = [0u8; DEFAULT_IV_SIZE];
+        iv[..U64_SIZE].copy_from_slice(&u64::to_ne_bytes(msg_seqno));
+        unsafe {
+            // Construct the request message header
+            (*self.req).hdr.hdr_sz = core::mem::size_of::<SnpGuestRequestMsgHdr>()
+                .try_into()
+                .unwrap();
+            (*self.req).hdr.algo = SNP_AEAD_AES_256_GCM;
+            (*self.req).hdr.hdr_version = HDR_VERSION;
+            (*self.req).hdr.msg_sz = core::mem::size_of_val(plaintext).try_into().unwrap();
+            (*self.req).hdr.msg_seqno = msg_seqno;
+            (*self.req).hdr.msg_type = msg_type;
+            (*self.req).hdr.msg_version = MSG_VERSION;
+            (*self.req).hdr.msg_vmpck = 0;
+            // Create new context
+            self.ctx_req = EVP_CIPHER_CTX_new();
+            if self.ctx_req.is_null() {
+                prints!("ERROR: Failed to get a new openssl CTX for encryption\n");
+                return false;
+            }
+            // Set encrypt operation
+            let mut ret: i32 = EVP_EncryptInit_ex(
+                self.ctx_req,
+                EVP_aes_256_gcm(),
+                ptr::null_mut(),
+                ptr::null(),
+                ptr::null(),
+            );
+            if ret != 1 {
+                prints!("ERROR: EVP_EncryptInit_ex failed ({})\n", ret);
+                return false;
+            }
+            // Provide key and iv lengths
+            ret = EVP_CIPHER_CTX_set_key_length(self.ctx_req, KEY_SIZE.try_into().unwrap());
+            if ret != 1 {
+                prints!(
+                    "ERROR: Failed to set the key length for encryption ({})\n",
+                    ret
+                );
+                return false;
+            }
+            ret = EVP_CIPHER_CTX_ctrl(
+                self.ctx_req,
+                EVP_CTRL_GCM_SET_IVLEN.try_into().unwrap(),
+                DEFAULT_IV_SIZE.try_into().unwrap(),
+                ptr::null_mut(),
+            );
+            if ret != 1 {
+                prints!("ERROR: EVP_CIPHER_CTX_ctrl failed ({})\n", ret);
+                return false;
+            }
+            // Provide key and iv
+            let svsm_secrets_va: VirtAddr = pgtable_pa_to_va(PhysAddr::new(svsm_secrets_page));
+            let svsm_secrets: *const SnpSecrets = svsm_secrets_va.as_ptr();
+            let key = *(&(*svsm_secrets).vmpck0 as *const _ as *const [u8; KEY_SIZE]);
```

Comment:
By creating an snpsecrets.rs file, these can be turned into a single function that retrieves the key.